### ISSUE TITLE
How to get cropped images from --save-crop in sequential order similar to original image?

### ISSUE THREAD

**Issue by ashmalvayani** [ultralytics/yolov5#7580]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I've run the inference of detecting multiple digits from an image, and the model is detecting all the numbers correctly; however, when I am cropping these digits and saving them using --save-crop, the outputs that I am getting are not in the order which is necessary for me to further link them to my next step.
the output should be 0 0 1 2 7, whereas the output is 1 7 0 2 0.
Is there any way to preserve the order? or anything through which I can rename the saved cropped images myself?

### Additional

![image](https://user-images.githubusercontent.com/56957881/165175503-efaca610-73c2-4f72-815c-1282058a75c5.png)
![image](https://user-images.githubusercontent.com/56957881/165175412-bfccc722-0fd1-4a54-8762-b5b04ba19a33.png)

**Comment by glenn-jocher** [comment]
@ashmalvayani 👋 Hello! Thanks for asking about **cropping results** with YOLOv5 🚀.  Cropping bounding box detections can be useful for training classification models on box contents for example. This feature was added in PR https://github.com/ultralytics/yolov5/pull/2827. You can crop detections using either detect.py or YOLOv5 PyTorch Hub. Crops are generally saved by order of confidence.

## detect.py
Crops will be saved under `runs/detect/exp/crops`, with a directory for each class detected.
```bash
python detect.py --save-crop
```
<img width="450" alt="Original" src="https://user-images.githubusercontent.com/26833433/139577918-65cc8f5d-741d-4a88-8753-0154013c624d.jpg">

<img width="1067" alt="Crop" src="https://user-images.githubusercontent.com/26833433/139577896-aa26b07c-466d-4057-9bfe-b01b56007587.png">

## YOLOv5 PyTorch Hub
Crops will be saved under `runs/detect/exp/crops` if `save=True`, and also returned as a dictionary with crops as numpy arrays.
```python
import torch

# Model
model = torch.hub.load('ultralytics/yolov5', 'yolov5s')  # or yolov5m, yolov5l, yolov5x, custom

# Images
img = 'https://ultralytics.com/images/zidane.jpg'  # or file, Path, PIL, OpenCV, numpy, list

# Inference
results = model(img)

# Results
crops = results.crop(save=True)  # or .show(), .save(), .print(), .pandas(), etc.
```

Good luck 🍀 and let us know if you have any other questions!

**Comment by Guttappa1238** [comment]
Hi @ashmalvayani If you are using torch hub, look at #36. For sorted results i.e. to sort license plate digit detection left-to-right (x-axis):

```
results = model(im)  # inference
results.pandas().xyxy[0].sort_values('xmin')  # sorted left-right
```
Hope it helps!

**Comment by Andy31798** [comment]
I have the same issue as @ashmalvayani. The saved cropped sequence is random. Any ways to let the save crop save from the most left x-axis to the most right x-axis? (Will edit some parts of the code in detect.py help?)

**Comment by glenn-jocher** [comment]
@Andy31798 not random. Sorted by confidence.

**Comment by Andy31798** [comment]
@glenn-jocher is there any parts in detect.py that I can modify so that the saved sequence will follow the order from the most left x-axis to the most right x-axis?

**Comment by glenn-jocher** [comment]
@Andy31798 the for loop that processes the results is here:
https://github.com/ultralytics/yolov5/blob/c4cb7c684ad6baf4081eb59ee1c27aa97bed4ebe/detect.py#L158-L172

**Comment by ashmalvayani** [comment]
@glenn-jocher can you recommend any changes in lines 158-172 to sort the results. I also found the same problem in this issue, can you tell the logic to sort via x-axis?
https://github.com/ultralytics/yolov5/issues/5573

**Comment by ashmalvayani** [comment]
**### Solved**

In the yolov5 directory, you'll have a detect.py file, there's a code line for NMS (Non-Max Suppression) in line 126 (depending upon the time you're cloning the GitHub repository it may vary and add the following code.

       # NMS
        pred = non_max_suppression(pred, conf_thres, iou_thres, classes, agnostic_nms, max_det=max_det)

        ## changing code here
        t = pred[0]
        pred = [torch.stack(sorted(t, key=lambda t: t[0]))]
        ## ending changed code here

        dt[2] += time_sync() - t3

This piece of code between ## solved my problem, however, it's giving me reversed sorted output so it's advised to use accordingly to your problem by removing sorted() for eg and many more.

**Comment by Ammar-Azman** [comment]
Hi @glenn-jocher , sorry to open this issue again. Based on this line `crops = results.crop(save=True)  # or .show(), .save(), .print(), .pandas(), etc.`, is there any option of method to save the bbox txt file?

**Comment by glenn-jocher** [comment]
Hello @Ammar-Azman! 

Thanks for sharing the solution that worked best for you. I'm sure it will be helpful to others who are having a similar issue. It's great to see that you were able to make that small modification to detect.py, which is the part of the code that processes detection results. You simply added a line of code that sorts the detections from left to right based on their x-axis coordinates.

It's important to know that sorting the results in a particular way depends on your specific use-case. In the solution you provided, you sorted the detections from left to right based on x-axis, which helped solve your issue. However, be aware that this solution will not work in all cases, as the properties of your input image and what you want to extract from it can vary.

Thanks for reaching out and please let us know if you have any other questions or issues!
